cmd/ocsp-responder/main.go

```diff
@@ -190,25 +210,163 @@ func (src *dbSource) Response(ctx context.Context, req *ocsp.Request) ([]byte, h
+		if err != nil {
+			src.log.AuditErrf("parsing OCSP response: %s", err)
+			ocspLookups.WithLabelValues("mysql_failed").Inc()
+			sourceUsed.WithLabelValues("error_returned").Inc()
+			return nil, nil, err
+		}
+		src.log.Debugf("returning ocsp from primary source: %v", helper.PrettyResponse(primaryParsed))
+		ocspLookups.WithLabelValues("mysql_success").Inc()
+		sourceUsed.WithLabelValues("mysql").Inc()
+		return primaryResult.bytes, header, nil
+	case secondaryResult := <-secondaryChan:
+		// If secondary returns first, wait for primary to return for
+		// comparison.
+		var primaryResult lookupResponse
+		// Listen for cancellation or timeout waiting for primary result.
+		select {
+		case <-ctx.Done():
+			err := fmt.Errorf("looking up OCSP response for serial: %s err: %w", serialString, ctx.Err())
+			src.log.Debugf(err.Error())
+			ocspLookups.WithLabelValues("canceled").Inc()
+			return nil, nil, err
+		case primaryResult = <-primaryChan:
+		}
+
+		if primaryResult.err != nil {
+			src.log.AuditErrf("Looking up OCSP response: %s", err)
+			ocspLookups.WithLabelValues("mysql_failed").Inc()
+			sourceUsed.WithLabelValues("error_returned").Inc()
+			return nil, nil, primaryResult.err
+		}
+		// Parse the OCSP bytes returned from the primary source to check
+		// status, expiration and other fields.
+		primaryParsed, err := ocsp.ParseResponse(primaryResult.bytes, nil)
+		if err != nil {
+			src.log.AuditErrf("parsing OCSP response: %s", err)
+			ocspLookups.WithLabelValues("mysql_failed").Inc()
+			sourceUsed.WithLabelValues("error_returned").Inc()
+			return nil, nil, err
+		}
+
+		secondaryParsed, err := ocsp.ParseResponse(secondaryResult.bytes, nil)
+		if err != nil {
+			src.log.Debugf("secondary OCSP lookup response error: %v", err)
+			ocspLookups.WithLabelValues("redis_failed").Inc()
+			sourceUsed.WithLabelValues("mysql").Inc()
+			return primaryResult.bytes, header, nil
+		}
+		if primaryParsed.Status != secondaryParsed.Status {
+			src.log.Err("primary ocsp source doesn't match secondary source, returning primary response")
+			ocspLookups.WithLabelValues("redis_mismatch").Inc()
+			sourceUsed.WithLabelValues("mysql").Inc()
+			return primaryResult.bytes, header, nil
+		}
+		src.log.Debugf("returning ocsp from secondary source: %v", helper.PrettyResponse(secondaryParsed))
+		ocspLookups.WithLabelValues("redis_success").Inc()
+		sourceUsed.WithLabelValues("redis").Inc()
+		return secondaryResult.bytes, header, nil
 	}
-	return certStatus.OCSPResponse, header, nil
+}
+
+type ocspLookup interface {
+	getResponse(context.Context, *ocsp.Request) chan lookupResponse
+}
+
+type redisReceiver struct {
+	rocspReader *rocsp.Client
+}
+type dbReceiver struct {
+	dbMap  dbSelector
+	filter *ocspFilter
+	log    blog.Logger
+}
+
+type lookupResponse struct {
+	bytes []byte
+	err   error
+}
+
+func (src dbReceiver) getResponse(ctx context.Context, req *ocsp.Request) chan lookupResponse {
```

Comment:
Each of these should have doccomments. In particular redisReceiver and dbReceiver should say they implement ocspLookup. BTW there's a cool trick to ensure at compile time they do implement that interface: https://splice.com/blog/golang-verify-type-implements-interface-compile-time/

Reply:
Added the comments and check. That is a great and obvious way to add an implementation check. Thanks!